docs/override_settings.md
## Індивідуальні налаштування сервера/маркетплейса ##

Для зручності можна створювати окремі файли для кожної групи налаштувань.

Як використовувати:

1. Створіть нову групу або візьміть потрібну групу налаштувань з /src/Config/marketplace.php
2. Помістіть зміни, згідно Правила оновлення налаштувань, в окремий файл (наприклад, для групи Marketplace в файл /assets/custom/Config/marketplace.php)

### Правила оновлення налаштувань ###

- Кількість файлів необмежена
- Файли обробляються по черзі, за алфавітним порядком
- Файли з підкресленням на початку назви ігноруються
- Кожен файл повертає масив інструкцій оновлення даних
- Масив не містить методів або анонімних функцій, використовуйте плейсхолдер '@_get(NAME)' для отримання значення всередині файлів налаштувань
- Для заміни елементу групи використовуйте макрос '_update' з переліком змін
- Для додавання елементу групи використовуйте макрос '_add'
- При поверненні асоціативного масиву без макросів буде виконано звичайне array_merge()
- Використовуйте '_match' пошуку елементів для макросів _add та _update.
- Для видалення елементу групи використовуйте макрос '_remove'
- Замість видалення (_delete) краще додайте до групм/елементу 'enabled' => false ("не використовувати").


Приклади:  
```
// оновити властивість 'enabled' = false для елемента Онлайн оплата (pt_online)
[
    '_match' => [
        'name' => 'pt_online'
    ],
    '_update' => [
        'enabled' => false,
    ]
],

// змінити label для Marketplace (вибір по uid )
[
    '_match' => [
        'uid' => '00000000-0001-0000-0000-000000000001'
    ],
    '_update' => [
        'label' => 'Тестовий маркетплейс (@_get(SERVER_NAME)'
    ]
],

// Деактивуємо елемент в pt (список Вид оплат) в Отримувачі коштів
[
    '_match' => [ // Шукаємо Payee по uid
        'uid' => '00000000-0003-0000-0000-0000000000011',
    ],
    '_update' => [
        'pt' => [
            [
                '_match' => [   // шукаємо pt по type
                    'type' => 'pt_online'
                ],
                '_update' => [
                    'enabled' => false
                ]
            ]
        ]
    ]
],

// Оновити label у елемента що має 'type' = 'email' та 'name' = '1'
[
    '_match' => [
        'type' => 'email',
        'name' => '1'
    ],
    '_update' => [
        'label' => 'Декларація 1'
    ]
],

// Прямме додавання елементу в Види оплат
[
    '_add' => [
        'Payment' => [
            [
                'name' => 'pt_bitcoin',
                'label' => 'Bitcoin'
            ]
        ]
    ]
],

// Приєднуємо новий тип оплати для Отримувача коштів (масив pt в Payee, пошук за uid)
[
    '_match' => [
        'uid' => '00000000-0003-0000-0000-000000000002', // Шукаємо конкретний uid
    ],
    '_add' => [
        'pt' => [
            [
                'type' => 'pt_bitcoint',
                'label' => 'Bitcoin (біток)',
                'provider' => 'Bitcoin'
            ]
        ]
    ]
],

// Видалити елемент з name pt_cashless з переліку Видів оплат
[
    '_delete' => [
        'Payment' => [
            [
                'name' => 'pt_cashless',
            ]
        ],
    ]
]
```

assets/plugins/combacart/src/Config/config.php
<?php

date_default_timezone_set('Europe/Kyiv');

const COMBAMODX_NAME = 'CombaCart';
const COMBAMODX_TDS = 'FS';
const COMBAMODX_FILE_VER = '20';
const COMBAMODX_VERSION = '2.6.' . COMBAMODX_FILE_VER . ' ' . COMBAMODX_TDS;

// термін актуальності кешованих даних кліента за замовчуванням
const COMBAMODX_CACHE_LIFETIME = 2592000; // 60*60*24* 30 діб

define("COMBAMODX_PATH_ROOT", dirname(__FILE__, 3));
const COMBAMODX_PATH_SRC = DIRECTORY_SEPARATOR . 'src';
const COMBAMODX_PATH_ASSETS = DIRECTORY_SEPARATOR . 'assets';
const COMBAMODX_PATH_CUSTOM = COMBAMODX_PATH_ASSETS . DIRECTORY_SEPARATOR . 'custom';
const COMBAMODX_PATH_THEMES = COMBAMODX_PATH_SRC . DIRECTORY_SEPARATOR . 'Themes';
const COMBAMODX_PATH_TEMPLATES = COMBAMODX_PATH_THEMES . DIRECTORY_SEPARATOR . 'templates';

const COMBAMODX_SESSION_NAME = 'SSUSER_';

const DB_SERVER_NAME = 'localhost';

// хост та доменне і'мя сайту,  https://something.com.ua something.com.ua
define("COMBAMODX_SERVER_HOST", 'https://' . getenv('HTTP_HOST'));
define("COMBAMODX_SERVER_NAME", getenv('SERVER_NAME'));

// мовний файл за замовчуванням
const COMBAMODX_LANGUAGE = 'uk';

// сторінка авторизації веб користувача
const COMBAMODX_PAGE_LOGIN = 'login';

// сторінка адміністрування замовленнями (за відсутності реєстрації на Comba сервері)
// можна змінити за неохідності
// http(s)://ваш_сайт/comba
const COMBAMODX_PAGE_COMBA = 'comba';

// сторінка Оформлення замовлення
const COMBAMODX_PAGE_CHECKOUT = 'checkout';
// сторінка перенаправлення після оформлення замовлення
const COMBAMODX_PAGE_TNX = 'tnx';
// тривалість актуальності даних на сторінці закінчення оформлення замовлення, сек
const COMBAMODX_PAGE_TNX_TIMEOUT = 259200; // 3 доби

// сторінка Відстеження замовлення
const COMBAMODX_PAGE_TRACKING = 't';
// сторінка Оплати замовлення
const COMBAMODX_PAGE_PAYMENT = 'p';
// сторінка callback Оплати замовлення (відповідь від платіжного сервісу)
const COMBAMODX_PAGE_PAYMENT_CALLBACK = 'ps';

// показувати Продавця на сторінках
const COMBAMODX_SELLER_SHOW = true;
// розділяти товари у Кошику по Продавцю на окремі Замовлення
const COMBAMODX_ORDER_SEPARATE_BY_SELLERS = true;

// максимальна кількість одного товару що додається у Кошик
const COMBAMODX_GOODS_MAX_QUANTITY = 99999;

// TV names
const GOODS_NAME = 'goods_name';
const GOODS_CODE = 'goods_code';
const GOODS_PRICE = 'goods_price';
const GOODS_PRICE_OLD = 'goods_price_old';
const GOODS_AVAIL = 'goods_avail';
const GOODS_WEIGHT = 'goods_weight';
const GOODS_ISONDEMAND = 'goods_isondemand';
const GOODS_ISNEWPRODUCT = 'goods_isnewproduct';
const GOODS_GOODS = 'goods_goods';
const GOODS_IMAGES = 'goods_images';
const GOODS_SELLER = 'goods_seller';
const GOODS_INBALANCES = 'goods_inbalances';


if (!function_exists('get3thAuth')) {
    /**
     * Basic config options
     * Override if exists config_private.php
     *
     * @param $provider
     * @return array
     */
    function get3thAuth($provider): array
    {
        $custom = array();
        if (file_exists(COMBAMODX_PATH_ROOT . COMBAMODX_PATH_CUSTOM . DIRECTORY_SEPARATOR . 'Config/config_custom.php')) {
            $custom = include COMBAMODX_PATH_ROOT . COMBAMODX_PATH_CUSTOM . DIRECTORY_SEPARATOR . 'Config/config_custom.php';
            if (!is_array($custom)) {
                $custom = array();
            }
        }

        // закоментуйте або поставте false для цього параметру якщо працюєте з повноцінним Comba сервером
        define("COMBAMODX_BUILDIN_SERVER", true);

        $default = array(
            /**
             * От саме тут налаштування Маркетплейсу для локального серверу
             **/
            'Marketplace' => ['uid' => '00000000-0001-0000-0000-000000000001'],
            /**
             * Ці дані використовуються лише для роботи з вбудованим локальним сервером!
             * Методи та властивості серверу \Comba\BuildInServer\Server можете змінювати за потреби, а саме:
             * marketplace()
             * sellers()
             * payee()
             **/


            /**
             * УВАГА!
             * Якщо ви здійснили реєстрацію магазину на Comba сервері
             * та бажаете повноцінно працювати, внесіть свої налаштування в
             * файл /assets/custom/Config/config_custom.php
             * Після цього налаштування для локального серверу та локального Маркетплейсу не використовуються.
             */
        );

        $auth = array_merge($default, $custom);

        return !empty($auth[$provider]) ? $auth[$provider] : array();
    }
}
